pkg/concurrencylimiter/limiter.go

```diff
@@ -0,0 +1,58 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package concurrencylimiter
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+)
+
+// Limiter limits concurrency of running tasks.
+type Limiter struct {
+	ch chan struct{}
+}
+
+// New creates a concurrency limiter.
+func New(limit int) *Limiter {
+	return &Limiter{
+		ch: make(chan struct{}, limit),
+	}
+}
+
+// Add adds one running task.
+// If max concurrency is reached, `Add` blocks until a running task finishes;
+// If context is cancelled, `Add` returnes an error immediately.
+func (l *Limiter) Add(ctx context.Context) error {
+	select {
+	case l.ch <- struct{}{}:
+		return nil
+	case <-ctx.Done():
+		return errors.New("context cancelled")
```

Comment:
I think we should use `ctx.Err()` here because it might be timeout caused by deadline.

Reply:
Done